HW4: Semenkin Eduard

Eduard585 opened the pull request.

Eduard585 committed:
Done task 4
Create ArrayQueueImpl
Create LinkedQueueImpl

Walingar commented on src/main/java/queue/impl/ArrayQueueImpl.java:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    private void shrink(int oldCapacity) {
+        int newCapacity = oldCapacity/CAPACITY_RATE;
```
spaces

Walingar commented on src/main/java/queue/impl/ArrayQueueImpl.java:
```diff
@@ -0,0 +1,111 @@
+
+    private void grow() {
+        int oldCapacity = this.elements.length;
+        int newCapacity = oldCapacity + (oldCapacity >> 1);
```
Делить на 2 лучше с помощью деления на 2)

Eduard585 replied:
![image](https://user-images.githubusercontent.com/33941067/80413124-a2dd9c00-88d7-11ea-9850-cc60e6760b1d.png)
Я подсмотрел в priorityQueue) Если честно не разобрался в вопросе) Почему в имплементации так?

Walingar replied:
Скорее всего, тому, кто писал реализацию, просто привычнее так писать деление на 2

Walingar replied:
Но лучше писать это `/ 2`, так будет проще понимать твой код большинству разработчиков

Walingar commented on src/main/java/queue/impl/ArrayQueueImpl.java:
```diff
@@ -0,0 +1,111 @@
+
+import java.util.*;
+
+public class ArrayQueueImpl<Integer> extends AbstractQueue<Integer> {
```
Ты тут сделал generic параметр, с названием Integer
Нужно сделать `public class ArrayQueueImpl extends AbstractQueue<Integer>`
Тогда не нужны будут касты

Walingar commented on src/main/java/queue/impl/ArrayQueueImpl.java:
```diff
@@ -0,0 +1,111 @@
+
+    private int head;
+    private int tail;
+    private Object[] elements;
```
И можно будет сделат массив Integer

Walingar commented on src/main/java/queue/impl/ArrayQueueImpl.java:
```diff
@@ -0,0 +1,111 @@
+            return null;
+        }
+
+        if (head > capacity - capacity / CAPACITY_RATE) {
```
Получается, после уменьшения размера, если мы будем делать poll/add то постоянно будем уменьшать, увеличивать размер массива, не хорошо. В таком случае делают так, чтобы после уменьшения размера массива, в нем оставалось элементов capacity / CAPACITY_RATE

Eduard585 replied:
@Walingar как я правильно понял, проблема в том, что мы начинаем туда сюда скакать после shrink т.к. head после shrink находится на границе определенной с помощью CAPACITY_RATE.  И решение в том, чтобы уменьшать массив не настолько сильно, чтобы head был подальше от границы? Т.е. чтобы tail - head  = newCapacity/CAPACITY_RATE?

Walingar replied:
> чтобы tail - head = newCapacity/CAPACITY_RATE

Да, все верно

Walingar commented on src/main/java/queue/impl/LinkedQueueImpl.java:
```diff
@@ -0,0 +1,87 @@
+
+    @Override
+    public Iterator<T> iterator() {
+        return null;
```
Забыл создать итератор)

Walingar commented on src/main/java/queue/impl/LinkedQueueImpl.java:
```diff
@@ -0,0 +1,87 @@
+        return new Node<>(value);
+    }
+
+    private class LinkedQueueIterator implements Iterator<T>{
```
Забыл имплементировать итератор )

Walingar commented on src/main/java/queue/impl/LinkedQueueImpl.java:
```diff
@@ -0,0 +1,87 @@
+        }
+    }
+
+    private static class Node<T>{
```
пробел

Walingar commented on src/main/java/queue/impl/LinkedQueueImpl.java:
```diff
@@ -0,0 +1,87 @@
+    private static class Node<T>{
+        private final T value;
+
+        public  Node<T> nextNode;
```
лишний пробел

Eduard585 committed:
Code review
Fix code style
Div by 2 instead of shifting
Fix return null instead of iterator

Walingar commented on src/main/java/queue/impl/ArrayQueueImpl.java:
```diff
@@ -0,0 +1,113 @@
+package queue.impl;
+
+import org.jetbrains.annotations.NotNull;
```
лишний импорт

Walingar commented on src/main/java/queue/impl/ArrayQueueImpl.java:
```diff
@@ -0,0 +1,113 @@
+            return null;
+        }
+
+        if (head > capacity - capacity / CAPACITY_RATE) {
```
Этот момент поправь и ок)

Walingar commented on src/main/java/queue/impl/ArrayQueueImpl.java:
```diff
@@ -0,0 +1,113 @@
+            try {
+                grow();
+            } catch (ArithmeticException ex) {
+                throw ex;
```
Непонятно, зачем ловить исключение, если оно сразу бросается и другие не ловятся
